### ISSUE TITLE
SAE/penalty robustness bugs (uncertainty covariance_mode; NuclearNorm HVP rank-cap panic; IBP underflow; weight*exp overflow; PCA-seed overflow)

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#742]
### ISSUE 1 — Full-uncertainty predictors ignore `covariance_mode`; never use corrected covariance, and mis-report provenance

**Title:** Predict full-uncertainty path ignores `PredictUncertaintyOptions.covariance_mode` — SEs always use conditional covariance and `covariance_corrected_used` is wrongly false

**Body:**
The full-uncertainty predict path records `covariance_mode_requested`
(`src/inference/predict/interval_policy.rs:556`) but never threads
`covariance_mode` into predictor SE construction
(`interval_policy.rs:526`: `transform.linear_state(input, fit, PredictPass::FullUncertainty)`).
Concrete predictors then compute SEs from the CONDITIONAL covariance only and never consult
the smoothing-corrected covariance:
- `src/inference/predict/mod.rs:2961-2966` (Bernoulli marginal-slope) via
  `require_posterior_mean_backend(...)`, which prefers `fit.beta_covariance()` and never
  `fit.beta_covariance_corrected()`.
- `src/inference/predict/mod.rs:3400-3407` (Gaussian location-scale) — same.
- `src/inference/predict/mod.rs:3736-3742` (Binomial location-scale) — uses `self.covariance`.
- `src/inference/predict/mod.rs:4328-4334` (Survival) — uses `self.covariance`.

`interval_policy.rs:557` then reports `covariance_corrected_used` via a default-false hook,
so when a caller requests the corrected/full covariance the returned SEs silently use the
narrower conditional covariance AND the provenance flag claims corrected was not used. Users
asking for `ConditionalPlusSmoothing` intervals get under-dispersed intervals with no error.

**Repro:** Gaussian location-scale fit, one mean coef,
`covariance_conditional = [[1,0],[0,0]]`, `covariance_corrected = [[9,0],[0,0]]`, one row
`x=[1]`, `predict_full_uncertainty(..., covariance_mode=ConditionalPlusSmoothingRequired)`
returns `eta_standard_error = 1.0` and `covariance_corrected_used = false`; correct is
`3.0` (or an error if corrected covariance is absent).

**Note:** Reviewer disproved the SAE shape-band candidate as correct: `sae_manifold.rs`
shape-band (~4552-4596) uses the right `(basis, channel)` indexing, one sqrt after
`phi Cov phi`, no missing cross-blocks for the per-atom/per-channel band.

---

### ISSUE 2 — NuclearNorm HVP rank cap uses right-Gram width instead of SVD rank → panic on wide decoder blocks

**Title:** NuclearNormPenalty HVP splits a tied zero-eigenvalue subspace for wide (M<p) blocks → panic

**Body:**
`value()`/`grad_target()` cap the active rank with `svd.singular.len()`
(`src/terms/analytic_penalties.rs:4036` and `:4051`), but `hvp` caps with the right-Gram
width `d` (`active_start = d.saturating_sub(self.rank_limit(d))`,
`src/terms/analytic_penalties.rs:3926`). For a wide block (`M < p`) where
`max_rank > min(n_eff, d)`, the value/grad treat the cap as a no-op (only `M` singular
values exist), but the HVP splits the tied zero-eigenvalue subspace of the `d×d` right Gram
and panics at `analytic_penalties.rs:3927-3937` ("max_rank splits a tied smoothed Gram
eigenvalue").

**Repro:** decoder block shape `3×10`, `nuclear_norm_max_rank=4`, full row-rank target.
Thin SVD has 3 singular values; value/grad use rank 3, HVP uses `d=10`, `active_start=6`,
splitting the seven equal `eps^2` Gram eigenvalues → panic. In SAE this is reachable via
`nuclear_norm_weight>0` with `nuclear_norm_max_rank` set and `M_k < p_out`.

**Fix sketch:** cap the HVP rank by `min(n_eff, d, svd_len)` consistently with value/grad
(use the same `rank_limit` against the true SVD length, not the Gram width).

---

### ISSUE 3 — Geometric IBP prior underflows to exact 0 for large K → atoms hard-masked with zero gradient (unrecoverable)

**Title:** IBP geometric prior underflows to 0.0 for large K, hard-masking atoms with zero assignment gradient

**Body:**
The geometric IBP prior accumulates `acc *= ratio` (`src/terms/sae_manifold.rs:7103`) where
`ratio = alpha/(alpha+1) < 1`, then multiplies the assignment and its gradient by this prior
(`out[i] = stable_logistic(...) * prior[i]` at `:7115`, `grad[i] = ... * prior[i]` at
`:7143`). For valid `(alpha, K)` the prior underflows to exact `0.0`, turning a soft
shrinkage prior into a HARD mask: those atoms get zero assignment AND zero logit gradient,
so they can never reactivate — contradicting the soft-decay contract.

**Repro:** `alpha=0.1`, `temperature=1`, `K=400`, zero logits. `ratio = 0.1/1.1`, and
`ratio.powi(320) == 0.0` in f64, so atoms with index >=~320 are unrecoverably masked.

**Fix sketch:** work in log-space for the geometric prior (accumulate `log(ratio)` and apply
via a stable log-sum-exp / logit shift) so the prior stays a finite soft weight, or floor
the prior at a tiny positive epsilon so the gradient path stays alive.

---

### ISSUE 4 — Learnable-weight penalties: unguarded `weight * rho.exp()` overflows to inf for large rho (systemic)

**Title:** Learnable penalty weights `weight * exp(rho)` overflow to inf / produce NaN for rho ≳ 709 (no clamp)

**Body:**
Many learnable-weight penalties resolve strength as `weight * rho[idx].exp()` with no clamp
or post-exp finiteness guard. For finite `rho ≳ 709` this overflows to `inf`, and downstream
`inf * 0.0` or `inf / inf` produces NaN in value/grad/Hessian, poisoning the solve. Sites
(non-exhaustive) in `src/terms/analytic_penalties.rs`: 1213, 1433, 1757, 1963, 2228, 2651,
3195, 3857, 4240, 4885, 5264, 6045, 6828 (DecoderIncoherence `resolved_weight`), 7113, 8308;
and in `src/terms/sae_manifold.rs`: `rho.log_lambda_sparse.exp()` at 7389/7458,
`rho.log_ard[..].exp()` at 3231/3658/3676. Also `JumpReLUPenalty` threshold
`thresholds[axis] * rho[axis].exp()` (`analytic_penalties.rs:2929`) → `inf` then
`inf * gate(0) = NaN` at ~2990. And `SparsityPenalty` REML smoothing `rho.exp()` UNDERflow
to `0.0` for `rho ≈ -1000` reintroduces `0/0` (`analytic_penalties.rs:2230,2275,...`).

These are mostly defense-in-depth (the optimizer's rho band is normally bounded), but a
caller-supplied or schedule-driven extreme rho can hit them. **Repro:** any learnable-weight
penalty with `rho=[710.0]` returns inf; e.g. softmax-entropy at `:1757` for nonzero entropy.

**Fix sketch:** centralize a `resolved_weight` helper that clamps the exponent (e.g. to
[-MAX_LOG, +MAX_LOG]) or floors/caps the resulting weight, and assert finiteness.

---

### ISSUE 5 — `sae_pca_seed_initial_coords` column-mean / SVD can overflow on huge finite Z (low priority)

**Title:** SAE PCA seed overflows column-mean accumulation to inf for huge finite Z before SVD

**Body:**
`sae_pca_seed_initial_coords` (`src/terms/sae_manifold.rs:11037-11048`) accumulates the
column mean as a plain running sum then centers and calls `.svd(true,true)`. Huge finite Z
(e.g. two rows of `1e308`) overflows the accumulation to `inf`; centered data is non-finite
before the SVD, producing a non-finite seed or SVD failure. Low priority (requires
pathological input) but a finiteness guard on Z at the SAE entry point would surface a clean
error. **Repro:** `sae_pca_seed` with two rows of `1e308`.

---

### Mismatch note (verify, possible doc bug) — DecoderIncoherence: docstring says `||B_j^T B_k||` but Rust penalizes `||B_j B_k^T||`

The Python docstring (`gamfit/_sae_manifold.py`, `decoder_incoherence_weight`) describes the
penalty as `||B_j.T @ B_k||_F^2`. The Rust `cross_gram`
(`src/terms/analytic_penalties.rs:6852-6872`) computes
`C[a,b] = Σ_o B_j[a,o]·B_k[b,o] = (B_j B_k^T)[a,b]` — i.e. it penalizes `||B_j B_k^T||_F^2`
(an `M_j×M_k` basis-space Gram), the OPPOSITE contraction from the docstring's
`B_j^T B_k` (a `p×p` feature-space Gram). value/grad/hvp are internally consistent with the
`B_j B_k^T` convention, so this is a documentation mismatch (or a spec mismatch) rather than
a numeric inconsistency — confirm which contraction is intended for "decoder column-space
incoherence" and align the docstring (the feature-space `B_j^T B_k` is the more natural
"column-space" reading, so the Rust contraction may itself be the bug). Pending codex
confirmation.

**Comment by SauersML** [comment]
Root-cause sweep update for the #742 bundle:

Several concrete bugs from this bundle are already fixed on current `main`:

- `d4b40fe0d` / `bd36ef5e3`: prediction uncertainty now propagates `PredictUncertaintyOptions.covariance_mode` through the generic full-uncertainty driver and predictor state. Predictors use `select_uncertainty_backend`, so `ConditionalPlusSmoothingPreferred/Required` can select `beta_covariance_corrected()` and report `covariance_corrected_used` honestly.
- `01e0bd2ac`: `NuclearNormPenalty` HVP now caps active rank by the true thin SVD rank, not the right-Gram width, so wide blocks such as `3 x 10` with `max_rank=4` no longer split the tied zero right-nullspace. Regression coverage includes the reported wide-block finite value/grad/HVP case.
- `fa52b277e`: IBP geometric assignment prior is accumulated in log space and floored before exponentiation, preserving a strictly-positive gradient path instead of hard-masking high-index atoms to exact zero. The same commit centralized learnable penalty strength resolution in log space so `weight * exp(rho)` does not overflow to `inf` or underflow to exact zero for finite rho. `tests/test_sae_manifold_ibp_prior_saturation.py` covers the high-K/low-alpha assignment saturation behavior.
- `37ea8889f`: added/kept runtime finiteness enforcement around learnable penalty weights.

Root causes, separated:

- Covariance path: the request mode was recorded as provenance but not used by predictor SE construction; predictors defaulted to conditional covariance and reported corrected covariance as unused.
- Nuclear norm: value/grad summed over thin SVD singular values while HVP used the wider right-Gram dimension, producing an inconsistent truncation boundary.
- IBP: value-space geometric multiplication made a soft shrinkage prior become an exact hard mask after f64 underflow.
- Learnable weights: exponentiating rho before multiplying made finite rho values produce `inf`, then downstream `inf * 0` / `inf / inf` NaNs.

Not closing yet because the PCA-seed huge-finite-input guard and DecoderIncoherence contraction/doc mismatch still need a separate decision or patch.

**Comment by SauersML** [comment]
Follow-up audit on the two items that were still open in the prior #742 update:

1. PCA seed huge-finite-input guard is now represented on `main`:
   - `01cf97fbc` changed `sae_pca_seed_initial_coords` to use stable column-mean accumulation and added finiteness guards.
   - `0f026b967` added a focused guard to pin the overflow case.

2. DecoderIncoherence contraction/doc mismatch is no longer present on current `main`:
   - `gamfit/_sae_manifold.py` now documents the penalty as `||B_j @ B_k.T||_F^2` for stored `(M_k, p_out)` decoder blocks.
   - `src/terms/analytic_penalties.rs` and `src/terms/penalties/decoder_incoherence.rs` document the same `B_j B_k^T` contraction.
   - Existing regression names also match the implemented basis-space cross-Gram semantics, e.g. `decoder_incoherence_heterogeneous_blocks_use_output_space_cross_gram` and `tests/test_sae_manifold_decoder_incoherence_crossgram.py`.

So the remaining source-level items listed in this bundle appear source-complete on `main`. I am not closing from this audit alone because this bundle is broad and still needs completed targeted runtime evidence across the affected guards; this comment only records that the previously named unresolved source gaps are no longer visible in current source.

**Comment by SauersML** [comment]
Fixed and verified on MSI. Relevant main commits include the robustness fixes/tests through `50f8818ef` and later build sync commits.

Targeted evidence:

```text
cargo test -q ibp_assignment_high_k_prior_keeps_positive_gradient_path --lib
1 passed

cargo test -q learnable_weights_stay_finite_at_extreme_rho --lib
1 passed

cargo test -q gaussian_location_scale_required_corrected_covariance_uses_corrected_backend --lib
1 passed

cargo test -q nuclear_norm --lib
6 passed

cargo test -q pca_seed_rejects_huge_finite_span_that_overflows_centering --lib
1 passed

cargo check -q -p gam-pyffi --lib
passed
```

These cover the filed covariance-mode, NuclearNorm HVP/rank-cap, IBP underflow/gradient path, learnable-weight overflow, and PCA-seed overflow regressions.
